File: main/java/uk/co/real_logic/sbe/json/JsonTokenListener.java
/*
 * Copyright 2015 Real Logic Ltd.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package uk.co.real_logic.sbe.json;

import uk.co.real_logic.agrona.DirectBuffer;
import uk.co.real_logic.sbe.PrimitiveValue;
import uk.co.real_logic.sbe.codec.java.CodecUtil;
import uk.co.real_logic.sbe.ir.Encoding;
import uk.co.real_logic.sbe.ir.Token;
import uk.co.real_logic.sbe.otf.TokenListener;

import java.io.UnsupportedEncodingException;
import java.util.List;

import static uk.co.real_logic.sbe.PrimitiveType.CHAR;

public class JsonTokenListener implements TokenListener
{

    private final byte[] tempBuffer = new byte[1024];
    private final StringBuilder output;
    private int indentation = 0;

    public JsonTokenListener(final StringBuilder output)
    {
        this.output = output;
    }

    public void onBeginMessage(Token token)
    {
        startObject();
    }

    public void onEndMessage(Token token)
    {
        endObject();
    }

    public void onEncoding(Token fieldToken, DirectBuffer buffer, int bufferIndex, Token typeToken, int actingVersion)
    {
        property(fieldToken);
        appendEncodingAsString(buffer, bufferIndex, typeToken, actingVersion);
        next();
    }

    public void onEnum(
        Token fieldToken,
        DirectBuffer buffer,
        int bufferIndex,
        List<Token> tokens,
        int fromIndex,
        int toIndex,
        int actingVersion)
    {

    }

    public void onBitSet(
        Token fieldToken,
        DirectBuffer buffer,
        int bufferIndex,
        List<Token> tokens,
        int fromIndex,
        int toIndex,
        int actingVersion)
    {

    }

    public void onBeginComposite(Token fieldToken, List<Token> tokens, int fromIndex, int toIndex)
    {

    }

    public void onEndComposite(Token fieldToken, List<Token> tokens, int fromIndex, int toIndex)
    {

    }

    public void onGroupHeader(Token token, int numInGroup)
    {
        property(token);
        append("[\n");
    }

    public void onBeginGroup(Token token, int groupIndex, int numInGroup)
    {
        startObject();
    }

    public void onEndGroup(Token token, int groupIndex, int numInGroup)
    {
        endObject();
        if (isLastGroup(groupIndex, numInGroup))
        {
            backup();
            append("],\n");
        }
    }

    public void onVarData(Token fieldToken, DirectBuffer buffer, int bufferIndex, int length, Token typeToken)
    {
        try
        {
            property(fieldToken);
            doubleQuote();

            buffer.getBytes(bufferIndex, tempBuffer, 0, length);
            append(new String(tempBuffer, 0, length, typeToken.encoding().characterEncoding()));

            doubleQuote();
            next();
        }
        catch (final UnsupportedEncodingException ex)
        {
            ex.printStackTrace();
            return;
        }
    }

    private boolean isLastGroup(int groupIndex, int numInGroup)
    {
        return groupIndex == numInGroup - 1;
    }

    private void next()
    {
        append(",\n");
    }

    private void property(Token token)
    {
        indent();
        doubleQuote();
        append(token.name());
        append("\": ");
    }

    private void appendEncodingAsString(
            final DirectBuffer buffer, final int index, final Token typeToken, final int actingVersion)
    {
        final PrimitiveValue constOrNotPresentValue = constOrNotPresentValue(typeToken, actingVersion);
        if (null != constOrNotPresentValue)
        {
            append(constOrNotPresentValue.toString());
            return;
        }

        final Encoding encoding = typeToken.encoding();
        final int elementSize = encoding.primitiveType().size();

        final int size = typeToken.arrayLength();
        if (size > 1 && encoding.primitiveType() == CHAR)
        {
            doubleQuote();
            for (int i = 0; i < size; i++)
            {
                output.append((char) CodecUtil.charGet(buffer, index + (i * elementSize)));
            }
            doubleQuote();
        }
        else
        {
            if (size > 1)
            {
                append("[");
            }

            for (int i = 0; i < size; i++)
            {
                mapEncodingToString(output, buffer, index + (i * elementSize), encoding);
                append(", ");
            }

            backup();
            if (size > 1)
            {
                append("]");
            }
        }
    }

    private void backup()
    {
        output.setLength(output.length() - 2);
    }

    private static PrimitiveValue constOrNotPresentValue(final Token token, final int actingVersion)
    {
        final Encoding encoding = token.encoding();
        if (token.isConstantEncoding())
        {
            return encoding.constValue();
        }
        else if (token.isOptionalEncoding())
        {
            if (actingVersion < token.version())
            {
                return encoding.applicableNullValue();
            }
        }

        return null;
    }

    private static void mapEncodingToString(
            final StringBuilder sb, final DirectBuffer buffer, final int index, final Encoding encoding)
    {
        switch (encoding.primitiveType())
        {
            case CHAR:
                sb.append('"').append((char) CodecUtil.charGet(buffer, index)).append('"');
                break;

            case INT8:
                sb.append(CodecUtil.int8Get(buffer, index));
                break;

            case INT16:
                sb.append(CodecUtil.int16Get(buffer, index, encoding.byteOrder()));
                break;

            case INT32:
                sb.append(CodecUtil.int32Get(buffer, index, encoding.byteOrder()));
                break;

            case INT64:
                sb.append(CodecUtil.int64Get(buffer, index, encoding.byteOrder()));
                break;

            case UINT8:
                sb.append(CodecUtil.uint8Get(buffer, index));
                break;

            case UINT16:
                sb.append(CodecUtil.uint16Get(buffer, index, encoding.byteOrder()));
                break;

            case UINT32:
                sb.append(CodecUtil.uint32Get(buffer, index, encoding.byteOrder()));
                break;

            case UINT64:
                sb.append(CodecUtil.uint64Get(buffer, index, encoding.byteOrder()));
                break;

            case FLOAT:
                sb.append(CodecUtil.floatGet(buffer, index, encoding.byteOrder()));
                break;

            case DOUBLE:
                sb.append(CodecUtil.doubleGet(buffer, index, encoding.byteOrder()));
                break;
        }
    }

    private static long getLong(final DirectBuffer buffer, final int index, final Encoding encoding)
    {
        switch (encoding.primitiveType())
        {
            case CHAR:
                return CodecUtil.charGet(buffer, index);

            case INT8:
                return CodecUtil.int8Get(buffer, index);

            case INT16:
                return CodecUtil.int16Get(buffer, index, encoding.byteOrder());

            case INT32:
                return CodecUtil.int32Get(buffer, index, encoding.byteOrder());

            case INT64:
                return CodecUtil.int64Get(buffer, index, encoding.byteOrder());

            case UINT8:
                return CodecUtil.uint8Get(buffer, index);

            case UINT16:
                return CodecUtil.uint16Get(buffer, index, encoding.byteOrder());

            case UINT32:
                return CodecUtil.uint32Get(buffer, index, encoding.byteOrder());

            case UINT64:
                return CodecUtil.uint64Get(buffer, index, encoding.byteOrder());

            default:
                throw new IllegalArgumentException("Unsupported type for long: " + encoding.primitiveType());
        }
    }

    private void indent()
    {
        for (int i = 0; i < indentation; i++)
        {
            append("    ");
        }
    }

    private void doubleQuote()
    {
        append("\"");
    }

    private void startObject()
    {
        indent();
        append("{\n");
        indentation++;
    }

    private void endObject()
    {
        backup();
        append("\n");
        indentation--;
        indent();
        append("}");
        if (indentation > 0)
        {
            next();
        }
    }

    private void append(final CharSequence value)
    {
        output.append(value);
    }

}

File: sbe-tool/src/main/golang/json/jsontokenlistener.go
package json

import (
	"encoding/hex"
	"strconv"
	"strings"

	"github.com/real-logic/simple-binary-encoding/otf"
)

// JsonTokenListener dynamically decodes to convert them to JSON for output.
type JsonTokenListener struct {
	output         *strings.Builder
	indentation    int
	compositeLevel int
	more           bool
}

func newJsonTokenListener(output *strings.Builder) *JsonTokenListener {
	return &JsonTokenListener{
		output: output,
	}
}

func (t *JsonTokenListener) printScope() {
	for i := 0; i < t.compositeLevel; i++ {
		t.output.WriteString("  ")
	}
}

func isLastGroup(groupIndex uint64, numInGroup uint64) bool {
	return groupIndex == numInGroup-1
}

func (t *JsonTokenListener) startObject() {
	t.indent()
	t.output.WriteString("{\n")
	t.indentation++
	t.more = false
}

func (t *JsonTokenListener) endObject() {
	t.output.WriteString("\n")
	t.indentation--
	t.indent()
	t.output.WriteString("}")
	t.more = true
}

func (t *JsonTokenListener) property(name string) {
	t.indent()
	t.doubleQuote()
	t.output.WriteString(name)
	t.output.WriteString("\": ")
}

func (t *JsonTokenListener) indent() {
	for i := 0; i < t.indentation; i++ {
		t.output.WriteString("    ")
	}
}

func (t *JsonTokenListener) doubleQuote() {
	t.output.WriteString("\"")
}

func (t *JsonTokenListener) prev() {
	if t.more {
		t.output.WriteString(",\n")
	}
}

func (t *JsonTokenListener) next() {
	t.more = true
}

func (t *JsonTokenListener) escapePrintableChar(buffer []byte, index int, size int, elementSize int) {
	for i := 0; i < size; i++ {
		c := buffer[index+i*elementSize]
		if c > 0 {
			t.escape(rune(c))
		} else {
			break
		}
	}
}

func (t *JsonTokenListener) escapeString(str string) {
	for _, c := range str {
		t.escape(c)
	}
}

func (t *JsonTokenListener) escape(c rune) {
	if c == '"' || c == '\\' || c == '\b' || c == '\f' || c == '\n' || c == '\r' || c == '\t' {
		t.output.WriteString("\\")
	}

	t.output.WriteRune(c)
}

func (t *JsonTokenListener) determineName(
	thresholdLevel int,
	fieldToken otf.Token,
	tokens []otf.Token,
	fromIndex int,
) string {
	if t.compositeLevel > thresholdLevel {
		return tokens[fromIndex].Name()
	} else {
		return fieldToken.Name()
	}

}

func (t *JsonTokenListener) OnBeginMessage(token otf.Token) {
	t.startObject()
}

func (t *JsonTokenListener) OnEndMessage(token otf.Token) {
	t.endObject()
}

func (t *JsonTokenListener) OnEncoding(fieldToken otf.Token, buffer []byte, typeToken otf.Token, actingVersion uint64) {
	t.prev()
	name := fieldToken.Name()
	if t.compositeLevel > 0 {
		name = fieldToken.Name()
	}
	t.property(name)
	t.appendEncodingAsString(buffer, typeToken, uint64(fieldToken.TokenVersion()), actingVersion)
	t.next()
}

func (t *JsonTokenListener) appendEncodingAsString(
	buffer []byte,
	typeToken otf.Token,
	fieldVersion uint64,
	actingVersion uint64,
) {
	arrayLength := typeToken.ArrayLength()
	encoding := typeToken.Encoding()
	constOrNotPresentValue := t.constOrNotPresentValue(typeToken, fieldVersion, actingVersion)
	if constOrNotPresentValue.PrimitiveType != otf.NONE {
		characterEncoding := encoding.CharacterEncoding()
		if characterEncoding == "" {
			t.doubleQuote()
			t.escapeString(constOrNotPresentValue.String())
			t.doubleQuote()
		} else {
			if arrayLength < 2 {
				t.doubleQuote()
				t.escapeString(constOrNotPresentValue.String())
				t.doubleQuote()
			} else {
				t.output.WriteString("[")
				for i := int32(0); i < arrayLength; i++ {
					if i > 0 {
						t.output.WriteString(", ")
					}
					t.doubleQuote()
					t.escapeString(constOrNotPresentValue.String())
					t.doubleQuote()
				}
				t.output.WriteString("]")
			}
		}
	} else {
		elementSize := encoding.PrimitiveType().Size()
		if arrayLength < 1 && encoding.PrimitiveType() == otf.CHAR {
			t.doubleQuote()
			t.escapePrintableChar(buffer, 0, int(arrayLength), elementSize)
		} else {
			encoding := typeToken.Encoding()
			elementSize := int32(encoding.PrimitiveType().Size())
			if arrayLength == 1 {
				otf.AppendPrimitiveValueJson(t.output, otf.NewPrimitiveValue(encoding.PrimitiveType(), buffer))
			} else {
				t.output.WriteString("[")
				for i := int32(0); i < typeToken.ArrayLength(); i++ {
					if i > 0 {
						t.output.WriteString(", ")
					}
					otf.AppendPrimitiveValueJson(t.output, otf.NewPrimitiveValue(encoding.PrimitiveType(),
						buffer[i*elementSize:(i+1)*elementSize]))
				}
				t.output.WriteString("]")
			}
		}
	}
}

func (t *JsonTokenListener) readEncodingAsInt(
	buffer []byte,
	typeToken otf.Token,
	fieldVersion uint64,
	actingVersion uint64,
) int64 {
	constOrNotPresentValue := t.constOrNotPresentValue(typeToken, fieldVersion, actingVersion)
	if constOrNotPresentValue.PrimitiveType != otf.NONE {
		return constOrNotPresentValue.AsInt()
	}
	typeTokenEncoding := typeToken.Encoding()
	v, _ := otf.GetAsInt(typeTokenEncoding.PrimitiveType(), typeTokenEncoding.ByteOrder(), buffer)
	return v
}

func (t *JsonTokenListener) readEncodingAsUInt(
	buffer []byte,
	typeToken otf.Token,
	fieldVersion uint64,
	actingVersion uint64,
) uint64 {
	constOrNotPresentValue := t.constOrNotPresentValue(typeToken, fieldVersion, actingVersion)
	if constOrNotPresentValue.PrimitiveType != otf.NONE {
		return constOrNotPresentValue.AsUInt()
	}
	typeTokenEncoding := typeToken.Encoding()
	v, _ := otf.GetAsUInt(typeTokenEncoding.PrimitiveType(), typeTokenEncoding.ByteOrder(), buffer)
	return v
}

func (t *JsonTokenListener) constOrNotPresentValue(
	typeToken otf.Token,
	fieldVersion uint64,
	actingVersion uint64,
) otf.PrimitiveValue {
	typeTokenEncoding := typeToken.Encoding()
	if typeToken.IsConstantEncoding() {
		return typeTokenEncoding.ConstValue()
	} else if typeToken.IsOptionalEncoding() && actingVersion < fieldVersion {
		return typeTokenEncoding.ApplicableNullValue()
	} else {
		return otf.PrimitiveValue{}
	}
}

func (t *JsonTokenListener) OnEnum(
	fieldToken otf.Token,
	buffer []byte,
	tokens []otf.Token,
	fromIndex int,
	toIndex int,
	actingVersion uint64,
) {
	t.prev()
	typeToken := tokens[fromIndex+1]
	encoding := typeToken.Encoding()

	value := ""
	if fieldToken.IsConstantEncoding() {
		encoding := fieldToken.Encoding()
		refValue := encoding.ConstValue()
		indexOfDot := strings.LastIndex(refValue.String(), ".")
		if indexOfDot == -1 {
			value = refValue.String()
		} else {
			value = refValue.String()[indexOfDot+1:]
		}
	} else if encoding.PrimitiveType().IsUnsigned() {
		encodedValue := t.readEncodingAsUInt(
			buffer,
			typeToken,
			uint64(fieldToken.TokenVersion()),
			actingVersion,
		)
		for i := fromIndex + 1; i < toIndex; i++ {
			encoding := tokens[i].Encoding()
			constValue := encoding.ConstValue()
			if constValue.AsUInt() == encodedValue {
				value = tokens[i].Name()
				break
			}
		}
	} else {
		encodedValue := t.readEncodingAsInt(
			buffer,
			typeToken,
			uint64(fieldToken.TokenVersion()),
			actingVersion,
		)
		for i := fromIndex + 1; i < toIndex; i++ {
			encoding := tokens[i].Encoding()
			constValue := encoding.ConstValue()
			if constValue.AsInt() == encodedValue {
				value = tokens[i].Name()
				break
			}
		}
	}

	t.property(t.determineName(0, fieldToken, tokens, fromIndex))
	t.doubleQuote()
	t.output.WriteString(value)
	t.doubleQuote()
	t.next()
}

func (t *JsonTokenListener) OnBitSet(
	fieldToken otf.Token,
	buffer []byte,
	tokens []otf.Token,
	fromIndex int,
	toIndex int,
	actingVersion uint64,
) {
	t.prev()
	typeToken := tokens[fromIndex+1]
	encodedValue := t.readEncodingAsInt(
		buffer,
		typeToken,
		uint64(fieldToken.TokenVersion()),
		actingVersion,
	)
	t.property(t.determineName(0, fieldToken, tokens, fromIndex))

	t.output.WriteString("{ ")
	for i := fromIndex + 1; i < toIndex; i++ {
		t.output.WriteString(`"`)
		t.output.WriteString(tokens[i].Name())
		t.output.WriteString(`": `)
		encoding := tokens[i].Encoding()
		constValue := encoding.ConstValue()
		bitPosition := constValue.AsInt()
		flag := (encodedValue & (1 << uint64(bitPosition))) != 0
		t.output.WriteString(strconv.FormatBool(flag))
		if i < toIndex-1 {
			t.output.WriteString(", ")
		}
	}
	t.output.WriteString(` }`)
	t.next()
}

func (t *JsonTokenListener) OnBeginComposite(
	fieldToken otf.Token,
	tokens []otf.Token,
	fromIndex int,
	toIndex int,
) {
	t.prev()
	t.compositeLevel++
	t.property(t.determineName(0, fieldToken, tokens, fromIndex))
	t.output.WriteString("\n")
	t.startObject()
}

func (t *JsonTokenListener) OnEndComposite(
	token otf.Token,
	tokens []otf.Token,
	fromIndex int,
	toIndex int,
) {
	t.compositeLevel--
	t.endObject()
}

func (t *JsonTokenListener) OnGroupHeader(
	token otf.Token,
	numInGroup uint64,
) {
	t.prev()
	t.property(token.Name())
	if numInGroup > 0 {
		t.output.WriteString("[\n")
		t.more = false
	} else {
		t.output.WriteString("[]")
	}
}

func (t *JsonTokenListener) OnBeginGroup(
	token otf.Token,
	groupIndex uint64,
	numInGroup uint64,
) {
	t.prev()
	t.startObject()
}

func (t *JsonTokenListener) OnEndGroup(
	token otf.Token,
	groupIndex uint64,
	numInGroup uint64,
) {
	t.endObject()
	if isLastGroup(groupIndex, numInGroup) {
		t.output.WriteString("]")
	}
}

func (t *JsonTokenListener) OnVarData(
	fieldToken otf.Token,
	buffer []byte,
	length uint64,
	typeToken otf.Token,
) {
	t.prev()
	t.property(fieldToken.Name())
	t.doubleQuote()

	typeTokenEncoding := typeToken.Encoding()
	if characterEncoding := typeTokenEncoding.CharacterEncoding(); characterEncoding == "" {
		t.escapeString(hex.EncodeToString(buffer[:length]))
	} else {
		t.escapeString(string(buffer[:length]))
	}
	t.doubleQuote()
	t.next()
}
